File: 2755-extra-characters-in-a-string/extra-characters-in-a-string.cpp
class Solution {
private:
    unordered_set<string> set;
    int dp[51];
    int solve(string &s,int start)
    {
        if(start==s.length())
        return 0;

        if(dp[start]!=-1)
        return dp[start];

        int result=1+solve(s,start+1);

        string curr="";
        for(int i=start;i<s.length();i++)
        {
            curr.push_back(s[i]);
            if(set.count(curr))
            {
                result=min(result,solve(s,i+1));
            }
        }
        return dp[start]=result;
    }

public:
    int minExtraChar(string s, vector<string>& dictionary) 
    {
        for(string word : dictionary)
        set.insert(word);

        memset(dp,-1,sizeof(dp));
        return solve(s,0);
    }
};

File: 2755-extra-characters-in-a-string/extra-characters-in-a-string.java
class Solution {
    public Solution()
    {
        set = new HashSet<>();
    }
    private Set<String> set;
    private Integer[] dp;
    private int solve(String s, int index)
    {
        if(index == s.length())
        return 0;

        if(dp[index] != null)
        return dp[index];

        int minChar = Integer.MAX_VALUE;
        StringBuilder sb = new StringBuilder();
        for(int i = index; i < s.length(); i++)
        {
            sb.append(s.charAt(i));

            int extraChar = 0;
            if(set.contains(sb.toString()))
            extraChar = solve(s, i + 1);
            else
            extraChar = i - index + 1 + solve(s, i + 1);

            minChar = Math.min(minChar, extraChar);
        }

        return dp[index] = minChar;
    }

    public int minExtraChar(String s, String[] dictionary) 
    {
        for(String word : dictionary)
        set.add(word);

        dp = new Integer[s.length()];
        return solve(s, 0);
    }
}